uuid: Add Universally Unique Identifier (UUID) utilities

sorru94 opened the pull request:
Add a basic implementation of UUID generating and manipulating functions.

Fixes: #77882

jukkar commented:
There are lot of uuid handling code in Bluetooth, should the BT use this library?

sorru94 commented:
> There are lot of uuid handling code in Bluetooth, should BT use this library?

I'm unfamiliar regarding the Bluetooth use of the UUID. I see it uses 16 and 32 bits partial UUIDs, which are not defined in the UUID [RFC](https://datatracker.ietf.org/doc/html/rfc9562).
We could add a couple of defines similar to the ones in the bluetooth driver ([`BT_UUID_INIT_128`](https://docs.zephyrproject.org/latest/doxygen/html/group__bt__uuid.html#ga1a840adf4bc06cf2cd5dbeb0c868374b) for example) to ease the transition.

jhedberg commented:
> > There are lot of uuid handling code in Bluetooth, should BT use this library?
> 
> I'm unfamiliar regarding the Bluetooth use of the UUID. I see it uses 16 and 32 bits partial UUIDs, which are not defined in [RFC 9562](https://datatracker.ietf.org/doc/html/rfc9562). We could add a couple of defines similar to the ones in the bluetooth driver ([`BT_UUID_INIT_128`](https://docs.zephyrproject.org/latest/doxygen/html/group__bt__uuid.html#ga1a840adf4bc06cf2cd5dbeb0c868374b) for example) to ease the transition.

AFAIK at least the 16- and 32-bit Bluetooth UUIDs don't follow any non-Bluetooth standard. The way the Bluetooth core specification deals with these, is that in practice everything is a 128-bit UUID, however UUIDs which are derived from the so-called Bluetooth Base UUID can be compressed into 16- and 32-bit formats (and these again expanded back into 128 bit format). I have a feeling we'll need to keep the Bluetooth UUID API separate, but some proper research into their compatibility with these RFC specs needs to be done. We might at most e.g. provide conversion functions to convert back and forth between bt_uuid_t and uuid_t.

jukkar commented:
@sorru94 could you check if there would be any non-BT code that could be changed to use this library?
At least in dhcpv6 https://github.com/zephyrproject-rtos/zephyr/blob/0237d375de5fe8912b9e4134f45f7774a155b977/subsys/net/lib/dhcpv6/dhcpv6_internal.h#L83 there is uuid handling.

Thalley commented on tests/lib/uuid/src/main.c:
```diff
@@ -0,0 +1,230 @@
+	/* Check version and variant fields */
+	zassert_equal(uuid_1[6U] >> 4U, 4U, "Generated UUID v4 contains an incorrect 'ver' field");
+	zassert_equal(uuid_1[8U] >> 6U, 2U, "Generated UUID v4 contains an incorrect 'var' field");
+	/* Check another generated UUID is different */
+	uuid_generate_v4(uuid_2);
+	zassert(memcmp(uuid_1, uuid_2, UUID_SIZE) != 0, "The two generated UUID v4 are equal.");
```
This is a bit of a funny test - 2 generated UUIDs can be the same, although very unlikely. We can probably keep it, but it actually fail with correct behavior

Thalley commented on tests/lib/uuid/src/main.c:
```diff
@@ -0,0 +1,230 @@
+	zassert_true(result == 0, "uuid_generate_v5 returned an error");
+	result = uuid_to_string(uuid, uuid_str, UUID_STR_LEN + 1);
+	zassert_true(result == 0, "uuid_from_string returned an error");
+	zassert_true(strcmp("2ed6657d-e927-568b-95e1-2665a8aea6a2", uuid_str) == 0,
+		     "uuid_str != 2ed6657d-e927-568b-95e1-2665a8aea6a2");
```
```suggestion
	zassert_str_equal("2ed6657d-e927-568b-95e1-2665a8aea6a2", uuid_str));
```

We actually have functions to compare strings (keeping it as is, is also fine)

Thalley commented on tests/lib/uuid/src/main.c:
```diff
@@ -0,0 +1,230 @@
+							  0x57, 0x5c, 0xaf, 0xe4, 0xce, 0x7f,
+							  0xc0, 0x3b, 0xcd, 0xc5};
+
+	int res = 0;
```
```suggestion
	int res;
```

Thalley commented on tests/lib/uuid/src/main.c:
```diff
@@ -0,0 +1,230 @@
+
+	res = uuid_from_string(first_uuid_v4_string, first_uuid_v4);
+	zassert_equal(0, res, "uuid_from_string returned an error");
+	for (size_t i = 0; i < 16; i++) {
+		zassert_equal(expected_first_uuid_v4_byte_array[i], first_uuid_v4[i],
+			      "first_uuid != from expected value");
+	}
```
Did you consider `zassert_mem_equal` for comparing arrays?

andyross commented on include/zephyr/sys/uuid.h:
```diff
@@ -0,0 +1,121 @@
+/** @brief Length of the UUID base64 URL and filename safe string representation. */
+#define UUID_BASE64URL_LEN 22
+
+/** @brief Binary representation of a UUID. */
+typedef uint8_t uuid_t[UUID_SIZE];
```
Driving by: I note carefully here that BT appears to demand a **little endian** UUID representation.

andyross commented on tests/lib/uuid/src/main.c:
```diff
@@ -0,0 +1,230 @@
+	uuid_t third_uuid_v4;
+	uuid_t first_uuid_v5;
+
+	uint8_t expected_first_uuid_v4_byte_array[16u] = {0x44, 0xb3, 0x5f, 0x73, 0xcf, 0xbd,
```
More drive-by: this is clearly byte-packing a **big endian** UUID.  The first element is a 32 bit dword and should be stored "73, 5f, b3, 44" on little endian systems.

andyross reviewed (commented):
Seems like there's an endianness collision between the existing code and the new library?  UUIDs basically made a huge mess of this with that text format, and almost nothing does it "correctly" (if there even is a "correct").  But 100% for sure we should be doing it **consistently**, and I don't think we are?

sorru94 commented on include/zephyr/sys/uuid.h:
```diff
@@ -0,0 +1,121 @@
+/** @brief Number of bytes in the binary representation of a UUID. */
+#define UUID_SIZE 16
+
+/** @brief Length of the UUID canonical string representation. */
+#define UUID_STR_LEN 36
+
+/** @brief Length of the UUID base64 string representation. */
+#define UUID_BASE64_LEN 24
+
+/** @brief Length of the UUID base64 URL and filename safe string representation. */
+#define UUID_BASE64URL_LEN 22
```
Yes, I should have mentioned it in the comments. 
I will make them all include the null terminator for coherency with the other library.

sorru94 commented on include/zephyr/sys/uuid.h:
```diff
@@ -0,0 +1,121 @@
+ * @retval 0 if the UUID has been correctly parsed and stored in @p out
+ * @retval -EINVAL if @p input is not acceptable
+ */
+int uuid_from_string(const char *input, uuid_t out);
```
Adding a method to initialize a UUID from a buffer can be useful. 
However, I would avoid creating a UUID from 16 and 32 buffers as it's specific to Bluetooth and not defined in the RFC.

sorru94 commented on lib/uuid/Kconfig:
```diff
@@ -0,0 +1,19 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config UUID
+	bool "Universally Unique Identifier (UUID) Support"
+	depends on BASE64
+	depends on ENTROPY_GENERATOR
+	help
+	  Enable use of UUID.
+
+config UUID_V5
+	bool "Universally Unique Identifier (UUID) Version 5 Support"
+	depends on UUID
+	depends on MBEDTLS
+	depends on MBEDTLS_MD
+	depends on MBEDTLS_SHA1
+	help
+	  Enable generation of UUID v5.
```
Yes, that makes sense.

sorru94 commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,211 @@
+	/* Content parsing */
+	size_t input_idx = 0U;
+	size_t out_idx = 0U;
+	const int uuid_str_base = 16;
```
It's only used in this function and should not be useful outside the library so I didn't want to make a file-wide define.

Thalley replied:
That was kind of my point. `UUID_STR_POSITION_FIRST_HYPHEN` is also only used in `should_be_hyphen`, but has made a `#define` instead. 

Ditto for `UUID_OFFSET_VERSION` only being used in `overwrite_uuid_version_and_variant`.

It doesn't need to be changed, but just curious about the inconsistency :)

sorru94 replied:
Oh, I didn't notice. 
I'm removing `uuid_str_base` as a side effect of using `hex2bin` as you suggested, so there should be no inconsistency.

Thalley commented on include/zephyr/sys/uuid.h:
```diff
@@ -0,0 +1,141 @@
+ * @param[in] data_size The size of the data buffer.
+ * @param[out] out The UUID where the result will be written.
+ *
+ * @retval 0 if the UUID has been correctly generated and stored in @p out
+ * @retval -EINVAL if @p input is not acceptable
+ * @retval -ENOMEM if memory allocation failed
+ * @retval -ENOTSUP if mbedTLS returned an unrecognized error
```
This applies to all functions added by this API

Thalley commented on include/zephyr/sys/uuid.h:
```diff
@@ -0,0 +1,141 @@
+ * @retval 0 if the UUID has been correctly parsed and stored in @p out
+ * @retval -EINVAL if @p data is null or @p data_size is too small
+ */
+int uuid_from_buffer(const uint8_t *data, size_t data_size, uuid_t out);
```
Doesn't `data_size` always need to be 16 octets? 

We could consider
```suggestion
int uuid_from_buffer(const uint8_t data[UUID_SIZE], uuid_t out);
```

Thalley commented on lib/uuid/Kconfig:
```diff
@@ -0,0 +1,36 @@
+config UUID_V4
+	bool "UUID version 4 generation support"
+	depends on UUID
+	depends on ENTROPY_GENERATOR
```
Is `ENTROPY_GENERATOR` required for this?

As far as I cant tell `uuid_generate_v4` only relies on `sys_rand_get` and I don't see any dependencies on that function.

sorru94 replied:
I think it is, without setting `ENTROPY_GENERATOR` I get a linking error when I build for most boards I tried. The error is an undefined reference to `z_impl_sys_rand_get`.

Thalley replied:
Okay, yeah that sort of makes sense. It's just an undocumented requirement of `sys_rand_get` it looks like :)

Thalley commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,231 @@
+	mbedtls_md_free(&ctx);
+
+	/* Store the computed SHA1 in the out struct */
+	for (size_t i = 0; i < UUID_SIZE; i++) {
```
```suggestion
	for (int i = 0; i < UUID_SIZE; i++) {
```

Minor; `UUID_SIZE` is currently defined as an `int`, and we should consider using the same type here. 

`UUID_SIZE` and the other macros can be defined to be `unsigned int` by appending `U` to the value, which would make this
```suggestion
	for (unsigned int i = 0; i < UUID_SIZE; i++) {
```
instead

Thalley commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,231 @@
+
+int uuid_from_buffer(const uint8_t *data, size_t data_size, uuid_t out)
+{
+	if ((data == NULL) || (data_size != UUID_SIZE)) {
+		return -EINVAL;
+	}
```
```suggestion
	if ((data == NULL) || (data_size != UUID_SIZE) || out == NULL) {
		return -EINVAL;
	}
```

Thalley commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,231 @@
+#endif
+
+#if defined(CONFIG_UUID_V5)
+int uuid_generate_v5(const uuid_t namespace, const void *data, size_t data_size, uuid_t out)
+{
+	int mbedtls_err = 0;
+	mbedtls_md_context_t ctx = {0};
+	const mbedtls_md_info_t *md_info = mbedtls_md_info_from_type(MBEDTLS_MD_SHA1);
+	const size_t sha_1_bytes = 20;
+	uint8_t sha_result[sha_1_bytes];
```
Missing input validation (several of the input values could be incorrect or `NULL`

Thalley commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,231 @@
+}
+
+#if defined(CONFIG_UUID_V4)
+void uuid_generate_v4(uuid_t out)
+{
```
`out` could be NULL, right?

Thalley commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,231 @@
+
+int uuid_from_string(const char *input, uuid_t out)
+{
+	if ((input == NULL) || (strlen(input) + 1 != UUID_STR_LEN)) {
+		return -EINVAL;
+	}
```
```suggestion
	if ((input == NULL) || (strlen(input) + 1 != UUID_STR_LEN) || out == NULL) {
		return -EINVAL;
	}
```

Thalley commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,231 @@
+
+int uuid_to_string(const uuid_t uuid, char *out, size_t out_size)
+{
+	if ((out == NULL) || (out_size < UUID_STR_LEN)) {
+		return -EINVAL;
+	}
```
```suggestion
	if (uuid == NULL || (out == NULL) || (out_size < UUID_STR_LEN)) {
		return -EINVAL;
	}
```

Thalley commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,231 @@
+{
+	size_t olen = 0;
+
+	if ((out == NULL) || (out_size < UUID_BASE64_LEN)) {
```
```suggestion
	if (uuid == NULL || (out == NULL) || (out_size < UUID_BASE64_LEN)) {
```

Thalley commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,231 @@
+	size_t olen = 0;
+	char uuid_base64[UUID_BASE64_LEN] = {0};
+
+	if ((out == NULL) || (out_size < UUID_BASE64URL_LEN)) {
```
```suggestion
	if (uuid == NULL || (out == NULL) || (out_size < UUID_BASE64URL_LEN)) {
```

andyross commented:
> However, I tried to follow as much as possible [RFC 9562](https://www.rfc-editor.org/rfc/rfc9562.html#section-4) for this library

Yeah... the problem is that "UUIDs" in practical code don't really and haven't ever followed that RFC, which is a post-hoc kind of cleanup thing.  Windows GUIDs are the obvious example of something that is always converted LE, and the code @Thalley posted above certainly implies the BT inherited the convention.

I'm don't have enough of a stake here to demand one or the other, I'm just saying (1) whatever code we commit needs to be **extremely** clear about the endianness convention and (2) it should match the conventions used by any existing in-tree users (or else support both, I guess, but... yuck)

Thalley commented on include/zephyr/sys/uuid.h:
```diff
@@ -0,0 +1,158 @@
+#endif
+
+/**
+ * @defgroup uuid UUID
+ * @ingroup utilities
+ * @{
```
Suggest to add `@since` and `@version 0.1.0` to indicate an experimental API, so that we can modify it more easily in the coming months in case we need to modify it. 

See https://docs.zephyrproject.org/latest/develop/api/overview.html#api-overview for more info

Thalley commented on include/zephyr/sys/uuid.h:
```diff
@@ -0,0 +1,158 @@
+ * @details This function will assume the input UUID to be in the Microsoft Component Object
+ * Model (COM) GUIDs little endian format.
+ *
+ * An UUID in the standard big-endian RFC9562 representation `00112233-4455-6677-8899-AABBCCDDEEFF`
+ * has the equivalent little-endian COM GUID representation `33221100-5544-7766-8899-AABBCCDDEEFF`.
```
Is this right? The last 2 values, `8899` and `AABBCCDDEEFF`, don't need to be swapped?

sorru94 replied:
The `8899` should be `9988` while the last one is correct. See the other related comment for more info.

Thalley replied:
> The `8899` should be `9988` while the last one is correct. See the other related comment for more info.

What other related comment? Can you share the link?

sorru94 replied:
Yes, if we follow the COM GUID representation the last two should not be swapped.

sorru94 replied:
Sorry, I deleted the comment as It was my mistake. I replied on the other comment and will push the changes shortly.

Thalley commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,276 @@
+#define UUID_V5_VERSION (5U)
+#define UUID_V5_VARIANT (2U)
+
+static void memcpy_swap(uint8_t *dst, const uint8_t *src, size_t length)
+{
+	__ASSERT(((src < dst && (src + length) <= dst) || (src > dst && (dst + length) <= src)),
+		 "Source and destination buffers must not overlap");
+
+	src += length - 1;
+
+	for (; length > 0; length--) {
+		*dst++ = *src--;
+	}
+}
```
Why not use `sys_memcpy_swap`?

Thalley commented on lib/uuid/uuid.c:
```diff
@@ -0,0 +1,276 @@
+	const size_t fifth_part_start = fourth_part_start + fourth_part_size;
+	const size_t fifth_part_size = 6U;
+
+	memcpy_swap(&out[first_part_start], &data[first_part_start], first_part_size);
+	memcpy_swap(&out[second_part_start], &data[second_part_start], second_part_size);
+	memcpy_swap(&out[third_part_start], &data[third_part_start], third_part_size);
+	memcpy_swap(&out[fourth_part_start], &data[fourth_part_start], fourth_part_size);
+	memcpy(&out[fifth_part_start], &data[fifth_part_start], fifth_part_size);
```
In the documentation for this function, only the first 3 parts are swapped, but this swaps the first 4. Why?

And why isn't the 5th swapped? (not familiar with the UUID spec)

sorru94 replied:
The issue with the endianness is that the UUID spec defines UUID only as big-endian.
The interpretation of little-endian UUID is not universal.
From my understanding here are our options for the UUID:
- Standard big-endian representation: `00112233-4455-6677-8899-AABBCCDDEEFF`
- little-endian representation inverting each block: `33221100-5544-7766-9988-FFEEDDCCBBAA`
- little-endian representation inverting the full UUID: `FFEEDDCC-BBAA-9988-7766-554433221100`
- little-endian representation for Microsoft COM GUIs representation: `33221100-5544-7766-8899-AABBCCDDEEFF`

I'm not sure which of those little-endian notations is used in the Bluetooth module. 
However, from my very superficial research, I think the last one is the more widely adopted.

Thalley replied:
I'll need to read up on the UUID spec. Endianess is always a bit tricky, but it usually comes down the the language and terminology used in the specs and whether the blocks are values or arrays of values.

sorru94 replied:
The blocks change based on the UUID version. For example, v5 is defined as follows: 
```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                           sha1_high                           |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|         sha1_high             |  ver  |      sha1_mid         |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|var|                       sha1_low                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                           sha1_low                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```
Where sha1_* are all components of the same SHA-1 hash.
While UUID v1 is defined as follows:
```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                           time_low                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|           time_mid            |  ver  |       time_high       |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|var|         clock_seq         |             node              |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                              node                             |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```
As a consequence, it might be difficult to find a generic conversion method.

Thalley commented on lib/uuid/Kconfig:
```diff
@@ -0,0 +1,36 @@
+
+menu "Universally Unique Identifier (UUID)"
+
+config UUID
+	bool "UUID support"
+	help
+	  Enable use of the UUID library.
+
+config UUID_V4
+	bool "UUID version 4 generation support"
+	depends on UUID
+	depends on ENTROPY_GENERATOR
+	help
+	  Enable generation of UUID v4.
+
+config UUID_V5
+	bool "UUID version 5 generation support"
+	depends on UUID
+	depends on MBEDTLS
+	depends on MBEDTLS_MD
+	depends on MBEDTLS_SHA1
+	help
+	  Enable generation of UUID v5.
+
+config UUID_BASE64
+	bool "UUID Base64 support"
+	depends on UUID
+	depends on BASE64
+	help
+	  Enable conversion functions to write UUIDs in base 64
+	  formats.
```
Similar to the experimental version in the header file, these should probably also `select EXPERIMENTAL` and add `[EXPERIMENTAL]` to their labels

Thalley reviewed (commented):
The implementation looks OK to me. Unsure how and if we can use this in BT due to endianess, so I suggest to make it experimental so that we can modify things until we figure that part out.

Thalley commented:
So after the RFC and looking more into this, I've noticed the following:

1. 4122, which is obsoleted by 9562, does seemingly not define any specific endianess
2. 9562 defines UUIDs to be stored as big-endian
3. MS documentation seemingly only refers to 4122
4. Haven't found any references to 4122 or 9562 from the BT specs
5. If we want to be 9562 compliant and treat all UUIDs a big-endian, then it's pretty each to support
6. If we want to support converting between little and big endian, then it's more tricky. Each version of UUID (version 1 to 8) have slightly different structures, and ideally each value in each version should be converted. As far as I can tell 9562 does not define how to represent those as strings, because it is big-endian. 
7. MS converts the first 3 values to little endian (https://devblogs.microsoft.com/oldnewthing/20220928-00/?p=107221) and then treats the 6-byte MAC address as a 6-octet array instead of a 6-octet value. MS seemingly also treats `time_high` and `ver` as a single value, instead of 12 and 4 bit values respectively. MS seemingly uses UUID version 1 (https://www.rfc-editor.org/rfc/rfc9562.html#name-uuid-version-1) for their GUIDs (UUIDs). The way that MS treats UUIDs is a complete mess :upside_down_face: 

To continue with this PR, I think we need to ask what it is we want to achieve. Do we want to library that implements 9562? Do we want a library that is compatible with MS COM? Do we want a library that is compatible with Bluetooth? 

The 9562 RFC has this nice little paragraph in the beginning that pretty much sums up the mess: 

> The use of UUIDs is extremely pervasive in computing. They comprise the core identifier infrastructure for many operating systems such as Microsoft Windows and applications such as the Mozilla Web browser; in many cases, **they can become exposed in many non-standard ways**.

The purpose of 9562 is to take RFC 4122 and properly standardize it.

sorru94 commented:
> So after the RFC and looking more into this, I've noticed the following:
> 
>     1. 4122, which is obsoleted by 9562, does seemingly not define any specific endianess
> 
>     2. 9562 defines UUIDs to be stored as big-endian
> 
>     3. MS documentation seemingly only refers to 4122
> 
>     4. Haven't found any references to 4122 or 9562 from the BT specs
> 
>     5. If we want to be 9562 compliant and treat all UUIDs a big-endian, then it's pretty each to support
> 
>     6. If we want to support converting between little and big endian, then it's more tricky. Each version of UUID (version 1 to 8) have slightly different structures, and ideally each value in each version should be converted. As far as I can tell 9562 does not define how to represent those as strings, because it is big-endian.
> 
>     7. MS converts the first 3 values to little endian (https://devblogs.microsoft.com/oldnewthing/20220928-00/?p=107221) and then treats the 6-byte MAC address as a 6-octet array instead of a 6-octet value. MS seemingly also treats `time_high` and `ver` as a single value, instead of 12 and 4 bit values respectively. MS seemingly uses UUID version 1 (https://www.rfc-editor.org/rfc/rfc9562.html#name-uuid-version-1) for their GUIDs (UUIDs). The way that MS treats UUIDs is a complete mess 🙃
> 
> 
> To continue with this PR, I think we need to ask what it is we want to achieve. Do we want to library that implements 9562? Do we want a library that is compatible with MS COM? Do we want a library that is compatible with Bluetooth?
> 
> The 9562 RFC has this nice little paragraph in the beginning that pretty much sums up the mess:
> 
> > The use of UUIDs is extremely pervasive in computing. They comprise the core identifier infrastructure for many operating systems such as Microsoft Windows and applications such as the Mozilla Web browser; in many cases, **they can become exposed in many non-standard ways**.
> 
> The purpose of 9562 is to take RFC 4122 and properly standardize it.

Yes, I agree that we need to clarify the scope of the library. 
My two cents are:
- This library could implement an RFC9562-compliant version of the UUIDs. 
- Libraries/drivers/modules using application-specific UUID variants could wrap around it or extend it following their requirements.

Thalley commented:
> Yes, I agree that we need to clarify the scope of the library. My two cents are:
> 
>     * This library could implement an RFC9562-compliant version of the UUIDs.
> 
>     * Libraries/drivers/modules using application-specific UUID variants could wrap around it or extend it following their requirements.

If this library is going to implement RFC 9562, then I don't expect it to be usable for other implementations that use a different RFC (e.g. 4122) or no RFC (like Bluetooth seemingly does). Given the 8 versions of UUID, we cannot easily transform UUIDs in a way that makes sense. 

Speaking of versions of UUID: As long as we keep everything big endian it's not a big deal, but if we do not, then we might need a `uuid_v1_t`, `uuid_v2_t`, etc. if we want to support little endianess in any capacity. 

That also means that a function like `uuid_from_buffer_le` will depend on the version, as each version have different structures and thus the way to transform from little to big endianess will differ for each version.

sorru94 commented:
> If this library is going to implement RFC 9562, then I don't expect it to be usable for other implementations that use a different RFC (e.g. 4122) or no RFC (like Bluetooth seemingly does). Given the 8 versions of UUID, we cannot easily transform UUIDs in a way that makes sense.
> 
> Speaking of versions of UUID: As long as we keep everything big endian it's not a big deal, but if we do not, then we might need a `uuid_v1_t`, `uuid_v2_t`, etc. if we want to support little endianess in any capacity.
> 
> That also means that a function like `uuid_from_buffer_le` will depend on the version, as each version have different structures and thus the way to transform from little to big endianess will differ for each version.

It's unclear to me if a method for endianess conversion can be discerned with certainty from the UUID version.
For sure the UUID version is one dependency. However, even within the same version there might be multiple representations possible.
The initial purpose for this library was to provide an UUIDs implementation following the latest RFC. 
Since the new RFC is quite exhaustive it makes it easy to create an "independent" library to parse and generate UUIDs.
The utility of this library wants to be for future development within the RTOS.
Making changes to make it compatible with the bluetooth is fine but should not come to a detriment to the usability of the library for new designs.

Thalley commented:
> It's unclear to me if a method for endianess conversion can be discerned with certainty from the UUID version.
> For sure the UUID version is one dependency. However, even within the same version there might be multiple representations possible.

As far as I can tell, there's only a single possible represenation: 
```
The formal definition of the UUID string representation is provided by the following ABNF [[RFC5234](https://datatracker.ietf.org/doc/html/rfc5234)]:

UUID     = 4hexOctet "-"
           2hexOctet "-"
           2hexOctet "-"
           2hexOctet "-"
           6hexOctet
hexOctet = HEXDIG HEXDIG
DIGIT    = %x30-39
HEXDIG   = DIGIT / "A" / "B" / "C" / "D" / "E" / "F"

Note that the alphabetic characters may be all uppercase, all lowercase, or mixed case, as per [Section 2.3](https://rfc-editor.org/rfc/rfc5234#section-2.3) of [[RFC5234](https://datatracker.ietf.org/doc/html/rfc5234)]. An example UUID using this textual representation from the above ABNF is shown in [Figure 1](https://datatracker.ietf.org/doc/html/rfc9562#sampleStringUUID).

f81d4fae-7dec-11d0-a765-00a0c91e6bf6
```

So regardless of version, it will always be `xxxxxxxx-xxxx-xxxx-xxxxxxxxxxxx` and since UUIDs are always Big endian (in RFC9562) then that represenation works with all versions. 

> The initial purpose for this library was to provide an UUIDs implementation following the latest RFC.
> Since the new RFC is quite exhaustive it makes it easy to create an "independent" library to parse and generate UUIDs.
> The utility of this library wants to be for future development within the RTOS.

That sounds good to me

> Making changes to make it compatible with the bluetooth is fine but should not come to a detriment to the usability of the library for new designs.

Given what I've learned, I don't see this being useful for BT. RFC9562 does not seem compatible with the Bluetooth RFC usage. 

I suggest that we remove the `uuid_from_buffer_le` again and rename `uuid_from_buffer_be` to `uuid_from_buffer` again, and as per the RFC treat all UUIDs as big endian, and then try to be full RFC9562 compliant. 

If we at any point in the future decide to work towards compatibility with BT or the MS version of UUIDs, then we can revisit such conversion functions.

Thalley commented:
@sorru94 please fix CI issues - I'll do another review after that

Thalley reviewed (dismissed):
No additional comments from me. 

It would have been nice to have a defined use case for this implementation (e.g. a sample or something), but not sure it's necessary. It's somewhat odd to add support for something without any uses, especially since this cannot be used with e.g. Bluetooth, but maybe the SOF module or something like that could use this in the future.

jhedberg commented:
@sorru94 wouldn't it make sense to also include a `uuid_copy(uuid_t dst, const uuid_t src)` API, since the underlying type is an array instead of e.g. a struct?

sorru94 commented:
> @sorru94 wouldn't it make sense to also include a `uuid_copy(uuid_t dst, const uuid_t src)` API, since the underlying type is an array instead of e.g. a struct?

Yes, I think it would be helpful. I will add it.

sorru94 commented:
> No additional comments from me.
> 
> It would have been nice to have a defined use case for this implementation (e.g. a sample or something), but not sure it's necessary. It's somewhat odd to add support for something without any uses, especially since this cannot be used with e.g. Bluetooth, but maybe the SOF module or something like that could use this in the future.

I agree that it's odd. However, the UUID is already in use in many places within the project. I hope this library will entice people to standardize its use.

sorru94 commented:
> > I agree that it's odd. However, the UUID is already in use in many places within the project. I hope this library will entice people to standardize its use.
> 
> Which project are you referring to here? The SOF?

I was referring to the Zephyr project. 
As @jukkar mentioned above `dhcpv6` uses some form of UUIDs. Furthermore, I could find some UUID in `fs/ext2.h`, `usb/bos.h`, and `drivers/tee.h`.
Hopefully, once this driver is available those modules could transition to using this one instead of ad hoc implementations.

Thalley commented:
> The one thing that concerns me a bit with the way you've defined the `uuid_t` type, is that it hides away the distinction of passing by value vs passing by reference. Maybe that's a good thing, but it's often nice to be able to see this very clearly in API usage by having pointer types vs value types, and using `&` to pass by reference, etc. Since the type that `uuid_t` is encapsulating is an array you end up in practice always passing by reference, which could come as a surprise, e.g. if a function expects to be able to modify an input parameter without it having any effects on the caller. I don't have a strong opinion either way, and there does seem to be benefits of the current solution, but I'm wondering if you did some careful comparison & analysis of defining the type as:
> 
> ```c
> typedef struct {
> 	uint8_t uuid[UUID_SIZE]
> } uuid_t;
> ```
> 
> vs the current:
> 
> ```c
> typedef uint8_t uuid_t[UUID_SIZE];
> ```

Doesn't our coding guidelines (or the Linux kernel guidelines) suggest not to `typedef` structs? 
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#typedefs

Another alternative would be to do
```c
struct uuid {
    uint8_t uuid[UUID_SIZE]
};
```

and then pass around `struct uuid`s which can be passed by value, by mutable reference or `const` reference (which I believe is the point of your suggestion here as well)

jhedberg commented:
@Thalley my impression is that Zephyr is much more relaxed to the usage of typedefs, but there should always be a some good justification for their usage.

jukkar commented:
> I was referring to the Zephyr project. 
> As @jukkar mentioned above `dhcpv6` uses some form of UUIDs. Furthermore, I could find some UUID in `fs/ext2.h`, `usb/bos.h`, and `drivers/tee.h`.
> Hopefully, once this driver is available those modules could transition to using this one instead of ad hoc implementations.
> 

I would suggest that at least one user of this api could be found and converted to use it in this pr. It is a bit weird to have an api in upstream that nobody uses.

sorru94 commented:
@Thalley and @jhedberg. As for your suggestion, I replaced the type definition with a struct. Let me know if it was what you intended.
@jukkar I tried to use this library in the ext2 library, just to have one that uses it.

Thalley commented:
@sorru94 Looks like CI tests are failing, please have a look

Thalley commented on include/zephyr/fs/ext2.h:
```diff
@@ -27,7 +29,7 @@ struct ext2_cfg {
 	uint32_t fs_size; /* Number of blocks that we want to take. */
 	uint32_t bytes_per_inode;
-	uint8_t uuid[16];
+	struct uuid uuid;
```
This is an API change, isn't it? 

If it is a stable API, then it needs to follow https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html#introducing-breaking-api-changes

An alternative which is a bit dirty would be to do
```suggestion
	union {
		uint8_t uuid[16];
		struct uuid uuid;
	};
```

So that it's not breaking. I'll leave this up to the maintainer(s) of the `fs`

sorru94 replied:
It's true, I'm not sure how much the uuid field is used in the current version of the ext2 module. It would be nice if the maintainers of the `fs` could help.
Another option for this change to be non-breaking would be to keep the `typedef uint8_t uuid_t[UUID_SIZE];` in uuid.h instead of the struct version.

Thalley replied:
@sorru94 I still think this needs to be considered an API change, which we probably should avoid or follow the procedure for breaking a stable API

Thalley commented on include/zephyr/sys/uuid.h:
```diff
@@ -0,0 +1,168 @@
+ * @retval 0 The UUID has been correctly copied in @p dst
+ * @retval -EINVAL @p dst is not acceptable
+ */
+int uuid_copy(struct uuid *dst, struct uuid src);
```
One thing I've noticed now is that you more often than not pass by value (e.g. `struct uuid src`) rather than by reference (`struct uuid *dst`). 

Is that something you have considered? Since the struct is 16 octets, that's quite a lot of data being passed around compared to a pointer (typically 4 octets). 

Should the API primarily use pass by reference instead of value?

sorru94 replied:
I usually pass by reference only when changes to the parameter are needed, i.e. when it's used as an output.
Do you know if copying 16 Bytes on each function call is an issue? Are there guidelines for this in the Zephyr project?

Thalley replied:
> Do you know if copying 16 Bytes on each function call is an issue? Are there guidelines for this in the Zephyr project?

I'm not aware of any guidelines, but pass by value would increase the stack usage, right? 

Personally I (almost) always pass by reference for structs (unless they are smaller than the size of a pointer).

Thalley commented on include/zephyr/sys/uuid.h:
```diff
@@ -0,0 +1,168 @@
+ * @retval 0 The UUID has been correctly copied in @p dst
+ * @retval -EINVAL @p dst is not acceptable
+ */
+int uuid_copy(struct uuid *dst, struct uuid src);
```
In most functions you also have `(src, dst)` (e.g. `uuid_from_buffer(const uint8_t data[UUID_SIZE], struct uuid *out);` , but this functions does `(dst, src)`. 

Consider making them consistent

Thalley commented on subsys/fs/ext2/ext2_format.c:
```diff
@@ -72,24 +67,24 @@ static void set_bitmap_bits(uint8_t *bitmap, uint32_t to_set)
+	in->i_mode = sys_cpu_to_le16(EXT2_DEF_DIR_MODE);
+	in->i_uid = 0;
+	in->i_size = sys_cpu_to_le32(nblocks * cfg->block_size);
+	in->i_atime = 0;
+	in->i_ctime = 0;
+	in->i_mtime = 0;
+	in->i_dtime = 0;
+	in->i_gid = 0;
+	in->i_blocks = sys_cpu_to_le32(nblocks * cfg->block_size / 512);
+	in->i_flags = 0;
+	in->i_osd1 = 0;
 	in->i_generation = 0;
-	in->i_file_acl   = 0;
-	in->i_dir_acl    = 0;
-	in->i_faddr      = 0;
+	in->i_file_acl = 0;
+	in->i_dir_acl = 0;
+	in->i_faddr = 0;
```
Looks like you've misconfigured your editor to format all lines in the file, and not only the lines you've changed. 

Please revert all format changes that are not changed

sorru94 commented:
> Sad to see this closed, @sorru94 do you think you can come back to this?

Sure, if someone reopens this I will address the comments.

pdgendt commented:
> Sure, if someone reopens this I will address the comments.

I'm not able to reopen it, maybe try to rebase and force push first? Other than that, I don't know.

sorru94 commented:
> @sorru94 if you reset the sorru94:add-uuid-utilities branch to commit [a93cdb1](https://github.com/zephyrproject-rtos/zephyr/commit/a93cdb14f25d4f32b12e18f7126aaadd841164d9) (SHA this PR was at when closed) we shall be able to re-open the PR

I did reset to [a93cdb1](https://github.com/zephyrproject-rtos/zephyr/commit/a93cdb14f25d4f32b12e18f7126aaadd841164d9). However, are you sure you didn't mean [18400cd](https://github.com/zephyrproject-rtos/zephyr/commit/18400cd58d5f1b8113e260045bb87f6f1f2d03b7)? I think that was the last commit in this PR.

pdgendt commented:
> I did reset to [a93cdb1](https://github.com/zephyrproject-rtos/zephyr/commit/a93cdb14f25d4f32b12e18f7126aaadd841164d9). However, are you sure you didn't mean [18400cd](https://github.com/zephyrproject-rtos/zephyr/commit/18400cd58d5f1b8113e260045bb87f6f1f2d03b7)? I think that was the last commit in this PR.

It can't be reopened now, so probably, yes.

pdgendt commented:
Reopened

sorru94 commented:
@Thalley 
Regarding [this](https://github.com/zephyrproject-rtos/zephyr/pull/77884#discussion_r1787634715) suggestion. I don't see how it would work. There is no apparent way to distinguish which of the two options is used. Accessing an array differs from accessing a struct.

It would have worked with my initial proposal of defining the uuid as a typedefinition of an array. But I have been pushed in the direction of using a struct. This makes it much harder to implement non-breaking changes to existing APIs.

I honestly do not know how to implement such a change in a non-breaking manner. Other suggestions are welcome.

Thalley commented:
> @Thalley Regarding [this](https://github.com/zephyrproject-rtos/zephyr/pull/77884#discussion_r1787634715) suggestion. I don't see how it would work. There is no apparent way to distinguish which of the two options is used. Accessing an array differs from accessing a struct.
> 
> It would have worked with my initial proposal of defining the uuid as a typedefinition of an array. But I have been pushed in the direction of using a struct. This makes it much harder to implement non-breaking changes to existing APIs.
> 
> I honestly do not know how to implement such a change in a non-breaking manner. Other suggestions are welcome.

IIRC the reason why you changed that, was to have a use of this new API, which makes sense. I wonder if there's a alternative thing we can modify to use the new API that does not require a change to a public API. Maybe a test or a sample? Alternatively, you could add a new sample that exercises some of the new functions you've added.

sorru94 commented:
> > @Thalley Regarding [this](https://github.com/zephyrproject-rtos/zephyr/pull/77884#discussion_r1787634715) suggestion. I don't see how it would work. There is no apparent way to distinguish which of the two options is used. Accessing an array differs from accessing a struct.
> > It would have worked with my initial proposal of defining the uuid as a typedefinition of an array. But I have been pushed in the direction of using a struct. This makes it much harder to implement non-breaking changes to existing APIs.
> > I honestly do not know how to implement such a change in a non-breaking manner. Other suggestions are welcome.
> 
> IIRC the reason why you changed that, was to have a use of this new API, which makes sense. I wonder if there's a alternative thing we can modify to use the new API that does not require a change to a public API. Maybe a test or a sample? Alternatively, you could add a new sample that exercises some of the new functions you've added.

I added a sample for the UUID module.
It seems the best way for this module to get merged and start being used subsequently.
As any change to a module currently using its ad hoc UUID implementation would break a public APIs.

Thalley commented on samples/subsys/uuid/README.rst:
```diff
@@ -0,0 +1,36 @@
+********
+
+This sample app demonstrates the use of the UUID utilities to generate and manipulate UUIDs
+accordingly to IETF rfc 9562.
```
```suggestion
accordingly to IETF RFC 9562.
```

Thalley commented:
@sorru94 looks good to me, but looks like there's some issue with the sample

sorru94 committed:
uuid: Add UUID utilities

Add UUID generation and parsing utilities compliant
with RFC9562.

Signed-off-by: Simone Orru <simone.orru@secomind.com>

sorru94 committed:
uuid: Add sample for UUID utilities

Add a sample that showcases the usqge of the
UUID utilities.

Signed-off-by: Simone Orru <simone.orru@secomind.com>

sorru94 committed:
uuid: Add documentation for UUID utilities

Include the UUID utilities in the Miscellaneous
documentation page.
The UUID is being placed in a new Identifier
APIs section.

Signed-off-by: Simone Orru <simone.orru@secomind.com>

sorru94 committed:
uuid: Add tests for UUID utilities

Add some tests for the UUID utilities.

Signed-off-by: Simone Orru <simone.orru@secomind.com>

Thalley commented on tests/lib/uuid/src/main.c:
```diff
@@ -0,0 +1,294 @@
+	zassert_equal(gen_uuid.val[8U] >> 6U, 2U,
+		      "Generated UUID v4 contains an incorrect 'var' field");
+}
+#else
+ZTEST(uuid, test_uuid_v4)
+{
+	ztest_test_skip();
+}
+#endif
```
```suggestion
#endif
```

It doesn't really make sense to have the `#else` here, does it? 

You should also be able to just add something like `Z_TEST_SKIP_IFNDEF(CONFIG_UUID_V4)` in the test itself

Thalley commented on lib/uuid/Kconfig:
```diff
@@ -0,0 +1,40 @@
+
+menu "Universally Unique Identifier (UUID)"
+
+config UUID
+	bool "UUID support [EXPERIMENTAL]"
+	select EXPERIMENTAL
+	help
+	  Enable use of the UUID library.
```
I realize that this comment is a bit last, but do we even need this at all? 

If none of the below is enabled, and none of the functions are called, the linker should remove the entire file by itself anyways.

sorru94 replied:
Removing this would make always available to the user functions such as `uuid_copy`, `uuid_from_buffer` and `uuid_from_string`. Since this is an experimental module I think it makes more sense to provide those functions only when explicitly using experimental features.

Thalley replied:
Sure. Since it's experimental, we could remove it when we it stops being experimental

pdgendt commented on samples/subsys/uuid/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2023 Nordic Semiconductor ASA
```
Fix copyright notice? If this is simply copied from another place, we should consider putting the generic copyright here.

```suggestion
# Copyright The Zephyr Project Contributors
```

sorru94 replied:
I ran through all the licensing headers and made them coherent.

pdgendt commented on include/zephyr/sys/uuid.h:
```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2024, SECO Mind Srl
```
There are some mixed copyright years/owners, are all of them accurate?

sorru94 replied:
I'll check them

kartben commented on doc/services/misc.rst:
```diff
@@ -34,3 +34,16 @@ this API is limited to creating the simplistic tokens needed to
+UUID
+====
+
+Universally Unique Identifiers (UUID), also known as Globally Unique
+IDentifiers (GUIDs) are an open, industry standard [RFC
+9562](https://tools.ietf.org/html/rfc9562) 128 bits long identifiers
+intended to guarantee uniqueness across space and time.
```
You are wrapping very aggressively here (72 characters?) --- 100 is fine/better :)

Also, you can be fancy with the Sphinx `:rfc:` role. Reworded + reformatted below:

```suggestion
Universally Unique Identifiers (UUID), also known as Globally Unique IDentifiers (GUIDs), are an
open industry standard (:rfc:`9562`) for 128-bit long identifiers intended to guarantee uniqueness
across space and time.
```

kartben reviewed (commented):
This is really great. Minor documentation comments, in case this goes through another round of updates (or if you want to address as a follow-up PR)

sorru94 commented:
> This is really great. Minor documentation comments, in case this goes through another round of updates (or if you want to address as a follow-up PR)

If there is no other large change, I would make this a follow-up PR.

tomi-font commented:
No entry has been added to the `MAINTAINERS.yml` file, so PRs touching these files have no one nor any label added to them. (Example: https://github.com/zephyrproject-rtos/zephyr/pull/90372)
Maybe something should be done about this?

tomi-font commented:
Again a PR that gets no reviewer assigned: https://github.com/zephyrproject-rtos/zephyr/pull/93565
We need to fix this.
@sorru94 would you be willing to step up as maintainer?

sorru94 commented:
> Again a PR that gets no reviewer assigned: #93565 We need to fix this. @sorru94 would you be willing to step up as maintainer?

@tomi-font I would be happy to help if needed.

tomi-font commented:
> @tomi-font I would be happy to help if needed.

https://discord.com/channels/720317445772017664/989195595572973648/1399669949861199965
